### ISSUE TITLE
Bootloader Plugin: efivar

### ISSUE THREAD

**Issue by Promaethius** [ostreedev/ostree#1903]
Are there any plans to implement efivar as a bootloader?
https://github.com/rhboot/efivar
For example, compiling the kernel with efi stubs and using secure boot direct to kernel.

**Comment by bauen1** [comment]
There's a number of different issues asking for different bootloader setup support: e.g.
- #1719 
- #2753 
- #2145 
- https://github.com/ostreedev/ostree/pull/2793

Ideally libostree only has to generate a BLS configuration and be done with it, however there are setups that would require special handling to integrate nicely with `ostree admin deploy`.

Would it be possible to provide a "bootloader" that simply runs hook scripts from the deployment ?
While far from ideal, it would make prototyping a lot easier, and less dirty than a fake `grub-mkconfig` script, like I'm doing for my UKI+SecureBoot setup.
And it would allow for exotic setups that have little to no chance of ever being upstreamed.
Ideally without chrooting, since that is complicated and if necessary could be done in those scripts.

**Comment by dbnicholson** [comment]
That's an interesting idea. Another way to potentially do it is to make the bootloader managers separate executables that are installed in `/usr/lib/ostree/bootloader` or similar. Then change the loading logic to match by basename. Then you could just drop in `/usr/lib/ostree/bootloader/mycustomthing` and then `ostree config set sysroot.bootloader mycustomthing` on the host.

Also, for your scenario I'd suggest explicitly setting the bootloader to `none` with `ostree config set sysroot.bootloader none`. Then all it does is create the boot loader config entries and you can do any hacks you need after `deploy`. We do this on Endless because we have grub patched to support BLS and do not want anything fiddling with `grub.cfg`.

**Comment by bauen1** [comment]
> That's an interesting idea. Another way to potentially do it is to make the bootloader managers separate executables that are installed in /usr/lib/ostree/bootloader or similar. Then change the loading logic to match by basename. Then you could just drop in /usr/lib/ostree/bootloader/mycustomthing and then ostree config set sysroot.bootloader mycustomthing on the host.

Wouldn't separate executables make things more difficult than necessary ?
And `aboot` seems to require special handling in `ostree-sysroot-deploy.c`.

While writing a `custom` bootloader manager that simply calls my scripts seems to be very simple, perhaps a 100 lines of code at most, I sadly don't have the time currently to setup an environment to implement and test it.

Just a couple more thoughts:
1. The script to call should probably be configurable e.g. by settting a `sysroot.bootloader_script` config on the repository.
2. The sysroot to use should be passed to the script, since there seems to be no other way of obtaining it, and it is necessary to make e.g. installers work nicely.
3. The `bootversion` needs to be passed along, it could be figured out by the script from looking at the `boot/loader` link, but the information is already easily available
4. Passing the `new_deployments` to the script seems unnecessary, since the kernel arguments, the kernel and initrd path can be obtained from parsing the BLS config files under `boot/loader.$bootversion/entries/`

I can't think of any scenario where chrooting before calling the script would be desirable.
Maybe when run from an installer, but I think implementing chroot in the script itself would be easier and more flexible.

In general I've noticed that there are quite a few different ways ostree-based chroots are setup, of course they are all slightly different.
For my setup alone I can count:

1. The installer setting up the installed deployment before the first boot
2. The initramfs (Granted, I'm currently using some shell script and not ostrees binary)
3. `ostree-bootloader-grub2.c`: https://github.com/ostreedev/ostree/blob/581a58067b6063f80434d2d136fdf421854151b0/src/libostree/ostree-bootloader-grub2.c#L275-L328 ; Except that is only used when not booted into ostree
4. Running `semodule --refresh` <https://github.com/ostreedev/ostree/blob/d49632faa6cbcda7b2891d5aab9f5a9a68348160/src/libostree/ostree-sysroot-deploy.c#L2911-L2964>

> Also, for your scenario I'd suggest explicitly setting the bootloader to none with ostree config set sysroot.bootloader none. Then all it does is create the boot loader config entries and you can do any hacks you need after deploy. We do this on Endless because we have grub patched to support BLS and do not want anything fiddling with grub.cfg.

Yes, I could, however I want to have `ostree admin deploy` work seamlessly.
My hack is already quite close to libostree just calling a script,  ideally I will be able to replace it with e.g. #2753 or libostree calling my scripts.

I've also found that I need to "wrap" grub-mkconfig in my modified Debian Installer (which is an even bigger hack), since the chroot provided by `ostree-bootloader-grub2.c` is lacking every important filesystem, not sure if that's because I'm doing something wrong, but it seems that <https://github.com/ostreedev/ostree/blob/490f515e189d1da4a0e04cc12b7a5e58e5a924dd/src/libostree/ostree-bootloader-grub2.c#L275-L328> expects `grub-mkconfig` to mount every filesystem it requires.

**Comment by cgwalters** [comment]
So I'm increasingly thinking that the bootloader backends all were a mistake, and the way things should work is via having separate systemd units that hook into `ostree-finalize-staged.service`.  The only case where this falls over is when we're operating on an alternative root - as in creating an initial deployment as part of a disk image.   But maybe for that we ask people to do it in external tooling.  

The appeal here is that it severs ostree from needing to know about specific bootloaders at all.  And systemd offers tons of controls and tooling for units - it's effectively already a plugin system today.  We'd just have to stabilize the existence of `ostree-finalize-staged.service`, but we already have things like greenboot hooking into it.

**Comment by bauen1** [comment]
So ostree would start the `ostree-finalize-staged.service` when not using `--stage` instead of doing it internally then ?

I don't particularly like this idea, because it would hide errors in systemd logs, but it works, so I could live with it.

Sorry, I'm pretty sure once I implement SecureBoot with my own PK, and TPM2.0 Attestation, I will manage to break a few more assumptions ostree makes about bootloaders :sweat_smile:

**Comment by ericcurtin** [comment]
aboot is similar to zipl backend in that, the main reason why having a backend is useful in this case is to flash boot data at the end of the deploy in the "post_bls_sync" stage. It's to execute final steps before next boot that an ostree commit simply cannot apply (as @cgwalters described in a different way).

There's actually a little demand for an Asahi Fedora Silverblue Remix now, if somebody actually implemented that, they might need similar hooks also, if you wanted to handle all kinds of upgrades in rpm-ostree, you also have the option of doing some upgrades outside of rpm-ostree like fwupd would do for example.

**Comment by jlebon** [comment]
> So I'm increasingly thinking that the bootloader backends all were a mistake, and the way things should work is via having separate systemd units that hook into `ostree-finalize-staged.service`. The only case where this falls over is when we're operating on an alternative root - as in creating an initial deployment as part of a disk image. But maybe for that we ask people to do it in external tooling.
> 
> The appeal here is that it severs ostree from needing to know about specific bootloaders at all. And systemd offers tons of controls and tooling for units - it's effectively already a plugin system today. We'd just have to stabilize the existence of `ostree-finalize-staged.service`, but we already have things like greenboot hooking into it.

Right, the issue with this I think which @bauen1 alluded to is that this only helps with the finalization flow at shutdown. I think whatever we come up with should work even if you direcly do `ostree admin finalize-staged` directly (or a more convincing argument where no staging is involved at all: `rpm-ostree cleanup`, which insta-writes the bootloader, or `rpm-ostree rollback` which could insta-write the bootloader).

**Comment by jlebon** [comment]
Definitely. It just feels like an awkward API vs just e.g. a dropin directory with executables (probably want both `/usr` and `/etc` variants).

**Comment by jlebon** [comment]
Here's a thorny example: you can have a staged deployment and a rollback deployment at the same time. If you just fire off `ostree-finalize-staged.service` on bootloader updates, `rpm-ostree cleanup -r` would also inadvertently finalize the staged deployment. So at the very least, I'd make it a separate unit instead, which `ostree-finalize-staged.service` would pull in.

But also, IMO I think systemd units which can be "re-activated" are just confusing. You almost always want `RemainAfterExit=yes`, but not in this case. Race conditions could also be an issue; ostree would probably want to wait until the service finishes before releasing the lock.
